[WGSL] Add support for texture_storage

tadeuzagallo opened the pull request:
#### 51203d9ac9e8fe78cc8eae7a1034619efa9bcea1
<pre>
[WGSL] Add support for texture_storage
<a href="https://bugs.webkit.org/show_bug.cgi?id=261465">https://bugs.webkit.org/show_bug.cgi?id=261465</a>
rdar://115363616

Reviewed by Dan Glastonbury.

Introduce a new TextureStorage type and new primitive types for texel format and
access mode.

* Source/WebGPU/WGSL/ConstantRewriter.cpp:
(WGSL::ConstantRewriter::materialize):
* Source/WebGPU/WGSL/Constraints.cpp:
(WGSL::satisfies):
(WGSL::satisfyOrPromote):
* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
(WGSL::bindingMemberForGlobal):
(WGSL::RewriteGlobalVariables::usesOverride):
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::emitPackedVector):
(WGSL::Metal::FunctionDefinitionWriter::visit):
* Source/WebGPU/WGSL/Overload.cpp:
(WTF::printInternal):
* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::TypeChecker):
(WGSL::TypeChecker::allocateTextureStorageConstructor):
* Source/WebGPU/WGSL/TypeStore.cpp:
(WGSL::VectorKey::encode const):
(WGSL::MatrixKey::encode const):
(WGSL::ArrayKey::encode const):
(WGSL::TextureKey::encode const):
(WGSL::TextureStorageKey::encode const):
(WGSL::ReferenceKey::extra const):
(WGSL::ReferenceKey::encode const):
(WGSL::TypeCache::find const):
(WGSL::TypeCache::insert):
(WGSL::TypeStore::TypeStore):
(WGSL::TypeStore::textureType):
(WGSL::TypeStore::textureStorageType):
(WGSL::VectorKey::extra const): Deleted.
(WGSL::MatrixKey::extra const): Deleted.
(WGSL::ArrayKey::extra const): Deleted.
(WGSL::TextureKey::extra const): Deleted.
(WGSL::TypeStore::TypeCache::find const): Deleted.
(WGSL::TypeStore::TypeCache::insert): Deleted.
* Source/WebGPU/WGSL/TypeStore.h:
(WGSL::TypeStore::accessModeType const):
(WGSL::TypeStore::texelFormatType const):
* Source/WebGPU/WGSL/Types.cpp:
(WGSL::Type::dump const):
(WGSL::Type::size const):
(WGSL::Type::alignment const):
* Source/WebGPU/WGSL/Types.h:

Canonical link: <a href="https://commits.webkit.org/267983@main">https://commits.webkit.org/267983@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2883483bf681926d115ee83dea212ecbc7c335b3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18270 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18604 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19181 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20109 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17099 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21901 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18757 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19035 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18491 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18700 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15908 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20992 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15923 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23157 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16942 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16832 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21047 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17397 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/14755 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16495 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4351 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20857 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17242 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebGPU/WGSL/TypeCheck.cpp:
```diff
@@ -1115,4 +1140,67 @@ void TypeChecker::allocateSimpleConstructor(ASCIILiteral name, TargetConstructor
+            auto& formatName = downcast<AST::IdentifierExpression>(type.arguments()[0]).identifier();
+            auto& accessName = downcast<AST::IdentifierExpression>(type.arguments()[1]).identifier();
+
+
```
Remove extra blank line?

djg reviewed (approved):
:shipit:

tadeuzagallo committed:
[WGSL] Add support for texture_storage
https://bugs.webkit.org/show_bug.cgi?id=261465
rdar://115363616

Reviewed by Dan Glastonbury.

Introduce a new TextureStorage type and new primitive types for texel format and
access mode.

* Source/WebGPU/WGSL/ConstantRewriter.cpp:
(WGSL::ConstantRewriter::materialize):
* Source/WebGPU/WGSL/Constraints.cpp:
(WGSL::satisfies):
(WGSL::satisfyOrPromote):
* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
(WGSL::bindingMemberForGlobal):
(WGSL::RewriteGlobalVariables::usesOverride):
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::emitPackedVector):
(WGSL::Metal::FunctionDefinitionWriter::visit):
* Source/WebGPU/WGSL/Overload.cpp:
(WTF::printInternal):
* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::TypeChecker):
(WGSL::TypeChecker::allocateTextureStorageConstructor):
* Source/WebGPU/WGSL/TypeStore.cpp:
(WGSL::VectorKey::encode const):
(WGSL::MatrixKey::encode const):
(WGSL::ArrayKey::encode const):
(WGSL::TextureKey::encode const):
(WGSL::TextureStorageKey::encode const):
(WGSL::ReferenceKey::extra const):
(WGSL::ReferenceKey::encode const):
(WGSL::TypeCache::find const):
(WGSL::TypeCache::insert):
(WGSL::TypeStore::TypeStore):
(WGSL::TypeStore::textureType):
(WGSL::TypeStore::textureStorageType):
(WGSL::VectorKey::extra const): Deleted.
(WGSL::MatrixKey::extra const): Deleted.
(WGSL::ArrayKey::extra const): Deleted.
(WGSL::TextureKey::extra const): Deleted.
(WGSL::TypeStore::TypeCache::find const): Deleted.
(WGSL::TypeStore::TypeCache::insert): Deleted.
* Source/WebGPU/WGSL/TypeStore.h:
(WGSL::TypeStore::accessModeType const):
(WGSL::TypeStore::texelFormatType const):
* Source/WebGPU/WGSL/Types.cpp:
(WGSL::Type::dump const):
(WGSL::Type::size const):
(WGSL::Type::alignment const):
* Source/WebGPU/WGSL/Types.h:

Canonical link: https://commits.webkit.org/267983@main
